rust/apps/zcash/src/pczt/check.rs

```diff
@@ -0,0 +1,341 @@
+use super::*;
+
+use orchard::{
+    commitment::ExtractedNoteCommitment,
+    keys::{FullViewingKey, Scope},
+    note::{Memo, Note, Nullifier, Rho},
+    note_ext::{calculate_note_commitment, calculate_nullifier},
+    Address,
+};
+use parse::decode_output_enc_ciphertext;
+use zcash_vendor::{
+    bip32::ChildNumber,
+    pczt,
+    ripemd::Ripemd160,
+    sha2::{Digest, Sha256},
+    zcash_primitives::legacy::{keys::AccountPubKey, Script, TransparentAddress},
+};
+
+pub fn check_pczt(
+    seed_fingerprint: &[u8; 32],
+    ufvk: &UnifiedFullViewingKey,
+    pczt: &Pczt,
+) -> Result<(), ZcashError> {
+    let xpub = ufvk.transparent().ok_or(ZcashError::InvalidDataError(
+        "transparent xpub is not present".to_string(),
+    ))?;
+    let orchard = ufvk.orchard().ok_or(ZcashError::InvalidDataError(
+        "orchard fvk is not present".to_string(),
+    ))?;
+    check_orchard(&seed_fingerprint, &orchard, &pczt.orchard())?;
+    check_transparent(&seed_fingerprint, &xpub, &pczt.transparent())?;
+    Ok(())
+}
+
+fn check_transparent(
+    seed_fingerprint: &[u8; 32],
+    xpub: &AccountPubKey,
+    bundle: &pczt::transparent::Bundle,
+) -> Result<(), ZcashError> {
+    bundle.inputs().iter().try_for_each(|input| {
+        check_transparent_input(seed_fingerprint, xpub, input)?;
+        Ok(())
+    })?;
+    bundle.outputs().iter().try_for_each(|output| {
+        check_transparent_output(seed_fingerprint, xpub, output)?;
+        Ok(())
+    })?;
+    Ok(())
+}
+
+fn check_transparent_input(
+    seed_fingerprint: &[u8; 32],
+    xpub: &AccountPubKey,
+    input: &pczt::transparent::Input,
+) -> Result<(), ZcashError> {
+    let script = Script(input.script_pubkey().clone());
+    match script.address() {
+        Some(TransparentAddress::PublicKeyHash(hash)) => {
+            let pubkey = input.bip32_derivation().keys().find(|pubkey| {
+                return hash[..] == Ripemd160::digest(Sha256::digest(pubkey))[..];
+            });
```

Comment:
Similarly, this check is incorrect:
- `hash` is the hash of the `script_pubkey`
- `input.bip32_derivation()` contains a map from the encoded public key to its derivation.

This follows PSBT semantics (from BIP 174) where the derivation map lists encoded keys; `bip32_derivation` is not a map from `script_pubkey` to derivation path. This is because a P2SH script may contain multiple such keys and require multiple signatures. A P2PKH script happens to only contain one, but for symmetry the PCZT format treats it the same way (like PSBT).